transaction/std/src/memo/gift_code_funding.rs

```diff
@@ -0,0 +1,310 @@
+        memo_data[0..Self::HASH_DATA_LEN]
+            .copy_from_slice(&tx_out_public_key_short_hash(tx_out_public_key));
+
+        // Put note into memo
+        let offset = Self::HASH_DATA_LEN;
+        memo_data[offset..(offset + note.len())].copy_from_slice(note.as_bytes());
+
+        Ok(Self { memo_data })
+    }
+
+    /// The Length of the TxOut hash
+    pub const HASH_DATA_LEN: usize = 4;
+
+    /// The length of the custom memo data
+    pub const MEMO_DATA_LEN: usize = 64;
+
+    /// Length of the utf-8 note
+    pub const NOTE_DATA_LEN: usize = 60;
+
+    /// Get the memo data
+    pub fn memo_data(&self) -> &[u8; Self::MEMO_DATA_LEN] {
+        &self.memo_data
+    }
+
+    /// Check if a given public key matches
+    pub fn public_key_matches(&self, tx_out_public_key: &RistrettoPublic) -> bool {
+        tx_out_public_key_short_hash(tx_out_public_key) == self.memo_data[0..Self::HASH_DATA_LEN]
+    }
+
+    /// Get funding note from memo
+    pub fn funding_note(&self) -> Result<&str, MemoError> {
+        let note = str::from_utf8(&self.memo_data[Self::HASH_DATA_LEN..])?;
+        if let Some(note) = note.split_once(char::from(0)) {
+            return Ok(note.0);
+        }
+        Ok(note)
+    }
+}
+
+// Compute first four bytes of TxOut hash
+fn tx_out_public_key_short_hash(
+    tx_out_public_key: &RistrettoPublic,
+) -> [u8; GiftCodeFundingMemo::HASH_DATA_LEN] {
+    let mut hasher = Blake2b512::new();
+    hasher.update("mc-gift-funding-tx-pub-key");
+    hasher.update(tx_out_public_key.as_ref().compress().as_bytes());
+    hasher.finalize().as_slice()[0..GiftCodeFundingMemo::HASH_DATA_LEN]
+        .try_into()
+        .unwrap()
+}
+
+impl From<&[u8; Self::MEMO_DATA_LEN]> for GiftCodeFundingMemo {
+    fn from(src: &[u8; Self::MEMO_DATA_LEN]) -> Self {
+        let mut memo_data = [0u8; Self::MEMO_DATA_LEN];
+        memo_data.copy_from_slice(src);
+        Self { memo_data }
+    }
+}
+
+impl From<GiftCodeFundingMemo> for [u8; GiftCodeFundingMemo::MEMO_DATA_LEN] {
+    fn from(src: GiftCodeFundingMemo) -> [u8; GiftCodeFundingMemo::MEMO_DATA_LEN] {
+        src.memo_data
+    }
+}
+
+impl_memo_type_conversions! { GiftCodeFundingMemo }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mc_util_from_random::FromRandom;
+    use rand::{rngs::StdRng, SeedableRng};
+
+    #[test]
+    fn test_gift_code_funding_memo_data_outputs_match() {
+        // Create memo from note and key
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let note = "Cash money MeowbleCoin for Kitty";
+        let key = RistrettoPublic::from_random(&mut rng);
+        let memo = GiftCodeFundingMemo::new(&key, note).unwrap();
+
+        // Check that the note is extracted properly
+        assert_eq!(memo.funding_note().unwrap(), note);
+
+        // Check that the public key can be verified
+        assert!(memo.public_key_matches(&key));
+    }
+
+    #[test]
+    fn test_gift_code_funding_memo_with_blank_note_is_ok() {
```

Comment:
Can we also have a test of what happens when the memo is the maximum length 56 bytes?

Reply:
nm you already did this :)